src/box/sql/select.c

```diff
@@ -270,6 +270,24 @@ sql_key_info_new_from_space_info(const struct sql_space_info *info)
 	return key_info;
 }
 
+/* Drop fields with type field_type_MAX from space info. */
+static void
+sql_space_info_normalize(struct sql_space_info *info, uint32_t drop_count)
+{
+	for (uint32_t j = 0; j < drop_count; ++j) {
+		uint32_t k = 0;
+		for (; k < info->field_count - j - 1; ++k) {
+			if (info->types[k] == field_type_MAX)
+				break;
+		}
+		for (; k < info->field_count - j - 1; ++k) {
+			info->types[k] = info->types[k + 1];
+			info->coll_ids[k] = info->coll_ids[k + 1];
+		}
```

Comment:
A canonical solution for removal of multiple items from an array is to open 2 iterators - src and dst. Src is copied into dst on each step. If src meets a condition, it is skipped but dst iterator is kept intact. That way you copy the array only once. Can be optimized if you firstly look for a first item to skip and if you don't find it, then you return. In your solution sometimes you do not copy anything indeed but otherwise some of the items can be copied multiples times.